ygen/protogen.go

```diff
@@ -88,21 +101,69 @@ message {{ .Name }} {
 	}
 )
 
+// writeProtoHeader outputs the header for a proto3 generated file. The package
+// name, base for the generated packages, files for which the protobuf was
+// generated, and the paths used for include are supplied as arguments. In
+// addition, whether path compression is being performed, and the calling
+// binary name are provided as arguments. Returns a string containing the
+// generated header code.
+func writeProtoHeader(pkg, basePkg, baseImportPath string, imports, yangFiles, includePaths []string, compressPaths bool, caller string) (string, error) {
+
+	if caller == "" {
+		caller = callerName()
+	}
+
+	if basePkg == "" {
+		basePkg = defaultBasePackageName
+	}
+
+	switch pkg {
+	case "":
+		pkg = basePkg
+	default:
+		pkg = fmt.Sprintf("%s.%s", basePkg, pkg)
+	}
+
+	in := struct {
+		PackageName      string
+		YANGFiles        []string
+		IncludePaths     []string
+		GeneratingBinary string
+		Imports          []string
+		BaseImportPath   string
+	}{
+		PackageName:      pkg,
+		YANGFiles:        yangFiles,
+		GeneratingBinary: caller,
+		IncludePaths:     includePaths,
+		Imports:          imports,
+		BaseImportPath:   baseImportPath,
+	}
+
+	var b bytes.Buffer
+	if err := protoTemplates["header"].Execute(&b, in); err != nil {
+		return "", err
+	}
+
+	return b.String(), nil
+}
+
 // writeProtoMsg generates a protobuf message for the *yangDirectory described by msg.
 // it uses the context of other messages to be generated (msgs), and the generator
 // state stored in state to determine names of other messages. compressPaths indicates
 // whether path compression should be enabled for the code generation. Returns a string
 // containing the name of the package that the message is within, a string containing
-// the generated code for the protobuf message, and any errors encountered during
+// the generated code for the protobuf message, a slice of strings containing the child
+// packages that are required by this message and any errors encountered during
 // proto generation.
-func writeProtoMsg(msg *yangDirectory, msgs map[string]*yangDirectory, state *genState, compressPaths bool) (string, string, []error) {
+func writeProtoMsg(msg *yangDirectory, msgs map[string]*yangDirectory, state *genState, compressPaths bool) (string, string, []string, []error) {
```

Comment:
Is this just proto3? If so it may be better  to call it writeProto3Msg. Just to be clear, I'm suggesting keeping type/function names version neutral if they support/can support all versions but adding version where a particular version is expected or emitted.
Another possibility is to have a enum parameter ProtoVersion2, ProtoVersion3 and passing that around.

Reply:
It is just proto3. I'll rename these functions to proto3.